RecoLocalTracker/SiPhase2VectorHitBuilder/src/VectorHitBuilderAlgorithm.cc

```diff
@@ -0,0 +1,547 @@
+          lpos_upp_corr = localParamsUpper[upperIterator].first.x();
+        } else if (localParamsUpper[upperIterator].first.x() < 0) {
+          lpos_low_corr = lparamsLow.first.x();
+          lpos_upp_corr = localParamsUpper[upperIterator].first.x() + std::abs(pC);
+        }
+      } else {
+        if (localParamsUpper[upperIterator].first.x() > 0) {
+          lpos_low_corr = lparamsLow.first.x();
+          lpos_upp_corr = localParamsUpper[upperIterator].first.x() - std::abs(pC);
+        } else if (localParamsUpper[upperIterator].first.x() < 0) {
+          lpos_low_corr = lparamsLow.first.x();
+          lpos_upp_corr = localParamsUpper[upperIterator].first.x() + std::abs(pC);
+        }
+      }
+
+      LogDebug("VectorHitBuilderAlgorithm") << " \t local pos upper corrected (x):" << lpos_upp_corr << std::endl;
+      LogDebug("VectorHitBuilderAlgorithm") << " \t local pos lower corrected (x):" << lpos_low_corr << std::endl;
+
+      //building my tolerance : 10*sigma
+      double delta = 10.0 * sqrt(lparamsLow.second.xx() + localParamsUpper[upperIterator].second.xx());
+      LogDebug("VectorHitBuilderAlgorithm") << " \t delta: " << delta << std::endl;
+
+      double width = lpos_low_corr - lpos_upp_corr;
+      LogDebug("VectorHitBuilderAlgorithm") << " \t width: " << width << std::endl;
+
+      unsigned int layerStack = theTkTopo->layer(stack->geographicalId());
+      if (stack->subDetector() == GeomDetEnumerators::SubDetector::P2OTB)
+        LogDebug("VectorHitBuilderAlgorithm") << " \t is barrel.    " << std::endl;
+      if (stack->subDetector() == GeomDetEnumerators::SubDetector::P2OTEC)
+        LogDebug("VectorHitBuilderAlgorithm") << " \t is endcap.    " << std::endl;
+      LogDebug("VectorHitBuilderAlgorithm") << " \t layer is : " << layerStack << std::endl;
+
+      float cut = 0.0;
+      if (stack->subDetector() == GeomDetEnumerators::SubDetector::P2OTB)
+        cut = barrelCut.at(layerStack);
+      if (stack->subDetector() == GeomDetEnumerators::SubDetector::P2OTEC)
+        cut = endcapCut.at(layerStack);
+      LogDebug("VectorHitBuilderAlgorithm") << " \t the cut is:" << cut << std::endl;
+
+      //old cut: indipendent from layer
+      //if( (lpos_upp_corr < lpos_low_corr + delta) &&
+      //    (lpos_upp_corr > lpos_low_corr - delta) ){
+      //new cut: dependent on layers
+      if (std::abs(width) < cut) {
+        LogDebug("VectorHitBuilderAlgorithm") << " accepting VH! " << std::endl;
+        VectorHit vh = buildVectorHit(stack, cluL, cluU);
+        //protection: the VH can also be empty!!
+        if (vh.isValid()) {
+          result.emplace_back(std::make_pair(vh, true));
+        }
+
+      } else {
+        LogDebug("VectorHitBuilderAlgorithm") << " rejecting VH: " << std::endl;
+        //storing vh rejected for combinatiorial studies
+        VectorHit vh = buildVectorHit(stack, cluL, cluU);
+        result.emplace_back(std::make_pair(vh, false));
```

Comment:
I think the check makes sense. 

The way it is done now is that both accepted and rejected VHs are used in the pattern recognition, but only accepted ones are used for seed building for pixelLess. So it makes sense to keep this distinction here.

Reply:
are the rejected passing any cut or does this include all `nLower*nUpper` possible pairings per stack?

Reply:
Right now this is the full combinatorics without any cut. There is a placeholder for a possible compatibility check before building, but right now it just returns true: https://github.com/cms-sw/cmssw/pull/31314/files/1e9919f9d68c5cf9d00a9befdb8c5f771d23abfe#diff-5e603863134d9e9c60a40d7211455b6bR107-R112

I think this is the reason why the efficiency for low pT tracks improves so much with the VHs. I think in the feature it would be worthwhile to define two sets of cuts, one for pattern recognition and one for building, but that involves a trade-off of speed vs low pT track efficiency that needs to be discussed.

Reply:
IIRC we have around 40 hits per module in the inner layer, does it mean we are creating some 10M hits per event? the VH purity in the inner then becomes just a few %. can tracking really be done well with so many fake hits? (I hope I got the math wrong, cause this would be really excessive, the profiling results would then be good only for hotspot detection, and the measured timing completely impractical of what a reasonable selection may have.)
do you have the stats of the number of hits, by chance?

Reply:
Just form 1 PU 200 event:
N accepted: 91278 N rejected 2058395
So your estimate is not too far off, unfortunately.